build: build fuzz tests by default

danben opened the pull request:
This fixes issue #19388. The changes are as follows:
  - Add a new flag to configure, --enable-fuzz-binary, which allows building test/fuzz/fuzz regardless of whether we are building to do actual fuzzing
  - Set -DPROVIDE_MAIN_FUNCTION whenever --enable-fuzz is no
  - Add the following libraries to FUZZ_SUITE_LD_COMMON:
    - LIBBITCOIN_WALLET
    - SQLLITE_LIBS
    - BDB_LIBS
    - if necessary, some or all of:
      - NATPMP_LIBS
      - MINIUPNPC_LIBS
      - LIBBITCOIN_ZMQ / ZMQ_LIBS



Fixes  #19388

practicalswift commented:
Concept ACK on building fuzz tests by default, but needs squashing and fixing of failing CI jobs :)

Haven't reviewed any code or build changes.

fanquake commented on configure.ac:
```diff
@@ -376,7 +376,7 @@ fi
   dnl First check if the compiler accepts flags. If an incompatible pair like
   dnl -fsanitize=address,thread is used here, this check will fail. This will also
-  dnl fail if a bad argument is passed, e.g. -fsanitize=undfeined
+  dnl fail if a bad argument is passed, e.g. -fsanitize=undefined
```
Isn't this typo an example of a bad argument?

danben replied:
Hi - apologies for making you read the code in this state. Initially I had opened the PR because I had a patch that was working under Linux and hadn't yet discovered the CI tests. Since then I've been pushing sporadically as a sanity check against the CI test behavior on my own machine. I haven't really worked with github before; is there a way to disable review on a PR (but not close the PR itself) until the code is ready?

I'm not sure how those commits got duplicated or how someone else's commits got pulled in. I believe it happened the last time I was attempting to rebase onto master. In any case I do intend to clean everything up once I can get all the tests to pass.

I'll restore the typo; at the time I hadn't realized that "undefined" was a valid argument.

I'm confused by your comment about building under windows, so I think I'm missing some fundamental point about the original issue. I had expected that as part of making the change the fuzz tests would now be built by default in more contexts (ie Windows, and specifically as an example the win64 CI test) and any compilation errors would need to be fixed as part of it. If that's not true, then as alternatives I can imagine (a) explicitly not building under non-compiling environments regardless of the value of ENABLE_FUZZ, or (b) ignoring the fact that the CI tests break (if that is even possible), but neither seemed that desirable? Anyway please let me know what behavior you'd like and I'll make the change.

danben replied:
(I suppose a third option would be to require someone building under e.g. windows to specify --disable-fuzz)

fanquake reviewed (commented):
I'd suggest splitting up this PR, because it seems like there are at least two separate changes here. i.e f1be9029e17ee702ddf19bbf6f08392ef4cdd8d6 "Allow building the fuzz tests under windows" shouldn't be related to whether or not we build the fuzz tests by default.

In any case, you need to rebase and fix up your branch, so that it only contains your changes. You can't have merge commits like 04f5799abfc683f91612fba41ba6e98bfa3b41a1. It looks like you've also got the same changes in multiple commits, i.e 622a2e7a5e47305574b04111ab8275f281158d12 and 0f2bdb1e26d262bb6d3d2950ea5ff81758a08f04. 

Please don't close and open a new PR.

sipa commented:
The standard "formula" for a generic swap is:

```c++
using std::swap;
swap(a, b);
```

danben commented:
> The standard "formula" for a generic swap is:
> 
> ```c++
> ...
> {
>     using std::swap;
>     swap(a, b);
> }
> ...
> ```
> 
> Not sure what the problem is, but maybe that's worth trying.

Thanks, I just hadn't looked carefully enough - it was a swap where one side was a vector location, and Element happened to be bool. Fingers crossed that this fixed it and I can clean up for review.

danben commented:
Well...almost there. I tried for a while to squash things down into a single commit, but I think Ive failed. The correct patch is at c7746f0 (except for the fact that I accidentally chopped off the first two lines of the commit message and now the linter is complaining), but I don't know how to get rid of the history (I used git rebase -i and manually squashed/dropped and this was the result). If someone could tell me how to fix this that would be much appreciated.

As far as the actual review, the substring match that I introduced feels kind of sketchy but I wasn't sure how else to accomplish what I wanted. The intent was to do some stuff only when libfuzzer was linked, which is what I thought the old check wanted to do as well but it still passed (that is, suppressed a main function etc) on e.g -fsanitize=thread. So I'm confused about what's going on there and would like feedback on that part in particular.

MarcoFalke commented:
```
git reset --soft 32b191fb66e644c690c94cbfdae6ddbc754769d7
git commit -m 'build: build fuzz tests by default'
# push ...

danben committed:
build: build fuzz tests by default.

This fixes issue #19388. The changes are as follows:
  - Add a new flag to configure, --enable-fuzz-binary, which allows building test/fuzz/fuzz regardless of whether we are building to do actual fuzzing
  - Set -DPROVIDE_MAIN_FUNCTION whenever --enable-fuzz is no
  - Add the following libraries to FUZZ_SUITE_LD_COMMON:
    - LIBBITCOIN_WALLET
    - SQLLITE_LIBS
    - BDB_LIBS
    - if necessary, some or all of:
      - NATPMP_LIBS
      - MINIUPNPC_LIBS
      - LIBBITCOIN_ZMQ / ZMQ_LIBS

sipa commented:
Code-wise, this looks pretty good, and in a short test it also seems to work.

I'm not entirely sure about the meaning of `--{enable,disable}-fuzz` that this introduces. There are really two distinct concepts now that are not exactly the same anymore:
* Whether or not you're building for some form of actual *fuzzing* (whether through -fsanitize=fuzzer, AFL, or something else). Doing any of these should disable all other targets.
* Whether or not you want to build the fuzz *code/binary* (which if fuzzing is enabled, will be what does the fuzzing; but otherwise is just a simple test binary that invokes the fuzz targets).

So I think this leads to problems if you'd want to fuzz with AFL. You'd need to specify (or get default) `--enable-fuzz`, and not enable `-fsanitize=fuzzer`, resulting in a situation where all the non-fuzz make targets aren't disabled. I don't think this is a big issue, and I'm perfectly fine with getting this merged as is, and iterate later. However, I think the following alternative may actually be less invasive change:
* Add a new `--{enable,disable}-fuzz-binary` option (default on) to indicate the need to build a ./src/test/fuzz/fuzz binary
* Leave `--{enable,disable}-fuzz` as option (default off) to indicate a fuzzing build (which disables all other make targets, and enables `--enable-fuzz-binary`)
* The existing detection for whether a main function is needed can stay.

danben commented:
@sipa Based on your comment I wasn't sure if I should make the change you suggested or wait for someone else to weigh in, but I wanted to make sure that I'm not dropping the ball on anything.

I agree that your suggestion would be easy to implement. If we don't want to add another flag (perhaps desirable from a usability perspective?), I could also just disable non-fuzz targets when building with AFL or Honggfuzz.

sipa commented:
@danben Yeah, up to you. Feel free to keep the PR as-is.

I don't think there is much convenience difference; in general nobody should touch --disable-fuzz-binary in general (except maybe we do so for release builds).

MarcoFalke commented on configure.ac:
```diff
@@ -1216,53 +1216,47 @@ dnl it would break GCC's #include_next.
 AC_DEFUN([SUPPRESS_WARNINGS],
          [$(echo $1 |${SED} -E -e 's/(^| )-I/\1-isystem /g' -e 's;-isystem /usr/include([/ ]|$);-I/usr/include\1;g')])
 
-dnl enable-fuzz should disable all other targets
```
On a second thought, this might be useful to keep. We don't want to waste resources compiling the wallet, util, cli, qt, bench, ... when the goal is to only fuzz.

danben replied:
Sorry, I'm confused by this. enable-fuzz refers to whether the fuzz targets should be built, which is now being done by default. It seems that we only want to disable the other targets when building with libfuzzer, afl or honggfuzz (the latter two of which still need to be addressed, which should be easy enough by inspecting CXX_FLAGS). Does this sound right to you?

MarcoFalke replied:
Before your change enable-fuzz means build the fuzz targets, but disable all other targets
After your change, it should keep that meaning, so that we don't waste compile resources. (I incorrectly assumed that it should go or change meaning).

After your change, the only thing that should change is that the fuzz target is always built (whether enable-fuzz or not).

There *could* be a disable-fuzz option that disables the fuzz target, but that can be done in a follow up. Unless you want to do it here right away.

sipa replied:
@MarcoFalke I think we're giving @danben conflicting instructions here?

Did you see https://github.com/bitcoin/bitcoin/pull/20936#issuecomment-766629435. I don't think we can just keep --{enable,disable}-fuzz with the same meaning.

danben replied:
I'm working on a patch that implement's @MarcoFalke's instructions as best I understand them just so that no matter how this gets resolved there will be a good place to begin wrapping things up.

The one thing that seemed a bit odd to me is this: if enable-fuzz used to mean "build the fuzz targets, but disable all other targets" and the expected change is that fuzz targets get built regardless of the value of enable-fuzz, then we end up with a flag whose name is enable-fuzz but all it does is disable building of non-fuzz targets.

Additionally, under these assumptions I don't think it's possible to punt on the flag that disables fuzz targets because without the ability to change code, some of the CI tests won't pass. But unfortunately this flag cannot be called disable-fuzz if we are saying that enable-fuzz has no effect on whether fuzz targets are built.

MarcoFalke replied:
In my comment I assumed that enable-fuzz and disable-fuzz are separate concepts (I don't know if configure can treat them as such). I used different words to describe the same that @sipa did in the previous comment. I don't care about the names of the configure arguments, as long as they achieve what we want :)

MarcoFalke commented on src/test/fuzz/fuzz.cpp:
```diff
@@ -72,7 +72,11 @@ extern "C" int LLVMFuzzerInitialize(int* argc, char*** argv)
+int main(int args, char **argv)
+#else
 __attribute__((weak)) int main(int argc, char** argv)
+#endif
```
I don't really like changing code in a build change. Maybe exclude the failing ci config via `--disable-fuzz` for now?

MarcoFalke commented on src/test/fuzz/netaddress.cpp:
```diff
@@ -9,7 +9,13 @@
+#include <winsock2.h>
+#else
 #include <netinet/in.h>
+#endif
```
I don't really like changing code in a build change. Maybe exclude the failing ci config via `--disable-fuzz` for now?

MarcoFalke commented on src/test/fuzz/string.cpp:
```diff
@@ -78,7 +78,9 @@ FUZZ_TARGET(string)
     (void)SanitizeString(random_string_1, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, 3));
+    #ifndef WIN32
     (void)ShellEscape(random_string_1);
+    #endif
```
I don't really like changing code in a build change. Maybe exclude the failing ci config via `--disable-fuzz` for now?

MarcoFalke commented on src/test/fuzz/util.h:
```diff
@@ -269,8 +269,15 @@ inline CNetAddr ConsumeNetAddr(FuzzedDataProvider& fuzzed_data_provider) noexcep
+        #else
         const in_addr v4_addr = {
             .s_addr = fuzzed_data_provider.ConsumeIntegral<uint32_t>()};
+        #endif
```
I don't really like changing code in a build change. Maybe exclude the failing ci config via `--disable-fuzz` for now?

MarcoFalke replied:
So this should probably be `--disable-fuzz-binary` in the windows ci (instead of my suggestion `--disable-fuzz`, which might not go well with how configure is supposed to work)

maflcko reviewed (commented):
I couldn't build a library with afl++, though the binaries did compile.

I expect this issue to "fix itself" once you address https://github.com/bitcoin/bitcoin/pull/20936#discussion_r566707448

```
root@9d90aa3c2826:/bitcoin# make
Making all in src
make[1]: Entering directory '/bitcoin/src'
make[2]: Entering directory '/bitcoin/src'
make[3]: Entering directory '/bitcoin'
make[3]: Leaving directory '/bitcoin'
  CXXLD    libbitcoinconsensus.la
afl-cc ++3.01a by Michal Zalewski, Laszlo Szekeres, Marc Heuse - mode: LLVM-PCGUARD
Using unoptimized trace-pc-guard, upgrade to llvm 10.0.1+ for enhanced version.
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:80: multiple definition of `__afl_area_ptr'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:80: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:81: multiple definition of `__afl_area_ptr_backup'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:81: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:79: multiple definition of `__afl_area_ptr_dummy'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:79: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_auto_early':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1045: multiple definition of `__afl_auto_early'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1045: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_auto_first':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1088: multiple definition of `__afl_auto_first'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1088: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_auto_init':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1033: multiple definition of `__afl_auto_init'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1033: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_auto_second':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1057: multiple definition of `__afl_auto_second'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1057: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:(.tbss+0x24): multiple definition of `__afl_cmp_counter'; /AFLplusplus/afl-compiler-rt.o:(.tbss+0x24): first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_coverage_discard':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1422: multiple definition of `__afl_coverage_discard'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1422: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_coverage_interesting':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1452: multiple definition of `__afl_coverage_interesting'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1452: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_coverage_off':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1400: multiple definition of `__afl_coverage_off'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1400: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_coverage_on':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1412: multiple definition of `__afl_coverage_on'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1412: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_coverage_discard':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1432: multiple definition of `__afl_coverage_skip'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1432: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:85: multiple definition of `__afl_fuzz_len'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:85: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_manual_init':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1004: multiple definition of `__afl_manual_init'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1004: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:88: multiple definition of `__afl_map_size'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:88: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_persistent_loop':
llvm_mode/instrumentation/afl-compiler-rt.o.c:945: multiple definition of `__afl_persistent_loop'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:945: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:(.tbss+0x20): multiple definition of `__afl_prev_ctx'; /AFLplusplus/afl-compiler-rt.o:(.tbss+0x20): first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:(.tbss+0x0): multiple definition of `__afl_prev_loc'; /AFLplusplus/afl-compiler-rt.o:(.tbss+0x0): first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:95: multiple definition of `__afl_selective_coverage_temp'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:95: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__afl_trace':
llvm_mode/instrumentation/afl-compiler-rt.o.c:135: multiple definition of `__afl_trace'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:135: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook1':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1212: multiple definition of `__cmplog_ins_hook1'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1212: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook2':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1235: multiple definition of `__cmplog_ins_hook2'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1235: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook4':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1256: multiple definition of `__cmplog_ins_hook4'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1256: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook8':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1277: multiple definition of `__cmplog_ins_hook8'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1277: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_rtn_hook':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1365: multiple definition of `__cmplog_rtn_hook'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1365: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook1':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1212: multiple definition of `__sanitizer_cov_trace_cmp1'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1212: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook2':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1235: multiple definition of `__sanitizer_cov_trace_cmp2'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1235: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook4':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1256: multiple definition of `__sanitizer_cov_trace_cmp4'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1256: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook8':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1277: multiple definition of `__sanitizer_cov_trace_cmp8'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1277: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook1':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1212: multiple definition of `__sanitizer_cov_trace_const_cmp1'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1212: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook2':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1235: multiple definition of `__sanitizer_cov_trace_const_cmp2'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1235: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook4':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1256: multiple definition of `__sanitizer_cov_trace_const_cmp4'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1256: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__cmplog_ins_hook8':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1277: multiple definition of `__sanitizer_cov_trace_const_cmp8'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1277: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__sanitizer_cov_trace_pc_guard':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1148: multiple definition of `__sanitizer_cov_trace_pc_guard'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1148: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__sanitizer_cov_trace_pc_guard_init':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1163: multiple definition of `__sanitizer_cov_trace_pc_guard_init'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1163: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `__sanitizer_cov_trace_switch':
llvm_mode/instrumentation/afl-compiler-rt.o.c:1326: multiple definition of `__sanitizer_cov_trace_switch'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:1326: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `at_exit':
llvm_mode/instrumentation/afl-compiler-rt.o.c:129: multiple definition of `at_exit'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:129: first defined here
/usr/bin/ld: /AFLplusplus/afl-compiler-rt.o: in function `send_forkserver_error':
llvm_mode/instrumentation/afl-compiler-rt.o.c:162: multiple definition of `send_forkserver_error'; /AFLplusplus/afl-compiler-rt.o:llvm_mode/instrumentation/afl-compiler-rt.o.c:162: first defined here
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:5482: libbitcoinconsensus.la] Error 1
make[2]: Leaving directory '/bitcoin/src'
make[1]: *** [Makefile:15061: all-recursive] Error 1
make[1]: Leaving directory '/bitcoin/src'
make: *** [Makefile:809: all-recursive] Error 1
```

danben commented on src/cuckoocache.h:
```diff
@@ -428,8 +428,10 @@ class cache
             // Can't std::swap a std::vector<bool>::reference and a bool&.
+            Element el = std::move(table[last_loc]);
+            table[last_loc] = std::move(e);
+            e = std::move(el);
```
Actually I'm a bit baffled by this one - the reason for this change was the instantiation of a bool cache in src/test/fuzz/cuckoocache.cpp, which looks like it would have been built whenever enable-fuzz was true. I don't understand how that ever compiled, but if you don't know the answer offhand I'll investigate.

MarcoFalke replied:
All compilers are broken (as in: They compile wrong code without error, or conversely compile correct code into an incorrect binary). But we should document this brokenness as best as we can, and fix incorrect code separate from feature changes or refactoring or build system changes. Mixing everything into one commit/patchset will make it harder to review and bisect later on.

danben commented:
Turned review back on. Two things to note:
- I have no idea what this appveyor failure is trying to tell me. It involves a file that I didn't touch.
- I left the new substring check for PROVIDE_MAIN_FUNCTION in order to be able to build some of the other sanitizer tests without disabling the fuzz binary. If anyone feels strongly that this should be reverted then I will.

practicalswift commented:
> * I have no idea what this appveyor failure is trying to tell me. It involves a file that I didn't touch.

Rebase on top of `master` and appveyor will be happily green again now that #21051 has been merged :)

MarcoFalke commented on configure.ac:
```diff
@@ -1221,6 +1227,14 @@ dnl it would break GCC's #include_next.
          [$(echo $1 |${SED} -E -e 's/(^| )-I/\1-isystem /g' -e 's;-isystem /usr/include([/ ]|$);-I/usr/include\1;g')])
 
+AC_MSG_CHECKING([whether test/fuzz/fuzz main function is needed])
+if [[[ $use_sanitizers == *"fuzzer"* ]]]; then
```
I am pretty sure this will break honggfuzz. Is there anything wrong with how this was checked previously?

MarcoFalke replied:
Ok, once you rebase I expect honggfuzz to be broken

maflcko reviewed (approved):
Concept ACK. Looks better now, except that it breaks honggfuzz

danben commented:
@MarcoFalke Prior to reopening I built and ran honggfuzz (as per doc/fuzzing.md) with no issue. Perhaps it's something to do with my system; how does it break for you?

Re: your question above, I had changed it because (for reasons that I still don't quite understand) the check would pass for me even when specifying e.g. --fsanitize=thread which caused several CI tests to fail now that the fuzz targets are built. That said, I would have expected the old check to do the right thing so I'm fairly confused by this.

MarcoFalke commented:
Oh, sorry. I was testing `master` :man_facepalming: , which means honggfuzz is broken in `master` and fixed in this pull? I'll need to test some more...

MarcoFalke commented:
>  That said, I would have expected the old check to do the right thing so I'm fairly confused by this.

I am incredibly confused by this as well. Tried to fix  in  #21080

MarcoFalke commented:
I think what you want now is:

```
if ENABLE_FUZZ:
 * Run MAIN_FUNCTION configure check
else:
 * PROVIDE_MAIN_FUNCTION=yes
fi

danben commented:
> I think what you want now is:
> 
> ```
> if ENABLE_FUZZ:
>  * Run MAIN_FUNCTION configure check
> else:
>  * PROVIDE_MAIN_FUNCTION=yes
> fi
> ```

Yep, got it changed locally and just testing everything now. Should be ready soon.

danben commented:
@MarcoFalke Strangely, now I'm seeing the same issue as in #19633. I reinstalled boost and tried a fresh build but it did not help. Pushing blindly for now.

MarcoFalke commented on configure.ac:
```diff
@@ -1250,21 +1250,21 @@ if test "x$enable_fuzz" = "xyes"; then
-      /* comment to remove the main function ...
-     ]],[[
-      */ int not_main() {
-     ]])])
```
Are the whitespace changes needed?

maflcko reviewed (approved):
ACK on this commit if the ci is green. Will test tomorrow

MarcoFalke commented:
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

MarcoFalke commented:
review ACK 32cbb06676e2957705d3ca7ff5710c9c3ff22c14 📭

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 32cbb06676e2957705d3ca7ff5710c9c3ff22c14 📭
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjAcQv/UflCiheoLDkcJSLBL3bLz9eoDY4DFaXXRstCZAmqnb4jdQYbW/uW60J4
B6Xhyx9mewauy+uKaspc+KOKVk3OSmGfnMa0ibIUuZ8w+SPUgKN/O2UeDfLEjqG5
e/wuYbOPNc9JI6xiPTRI6XRCtUnWLSPK4x9lymqbUS3lNtBQ+JupW+fxTQK1+lC2
xb//K9l9wnMoq8ZyOmfVbprzrDgyfSpEo3jTdeVlo5FraE7WQrgCSR0uix2QxZfp
DX7FOJrG3mld8PhR4AMQ0pZgJv4fDUn3fqMEg36/Aaz/SRUXpSCAZdZ1vNU3xKjv
CyyPmLAaAhDfIrKyPy7KLee6VpeczM4klMdWKI8bzVLlPs8kUGJPvc9uaaE7CUua
ValfGAXpN8tzdhmss8qvQyd50SOjnBhhoRcpVTVfSMttEoYOLXoMSfHhmMK7JC9h
YPqXcER367uTsINlUq3yMxJkYK6kkT0D/ki6doTjppy9YAzFCZin2zPpj/X6Zvgk
9O61+Nfq
=eqEg
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6ff121421f769a4f3402980038ef6e89d55175dfce07728a3f9adfac23f10030  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086ff121421f769a4f3402980038ef6e89d55175dfce07728a3f9adfac23f10030f01028607e4271190fed30759fb084f574d808fff01026bebcde03c554ccfdae8b8ba2d0636d08f104601e4b37f00860137cb0439e45340083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010454b5e659c684815d24fe3bc0e0f483508f104601e4b38f008b9f2fa72989e22cf0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a8074d29215ae5b642ddbf2e3e42f53d08f104601e4b37f008c90f4edc344895f30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010b606212a0ca722d9123462b60e74fdd208f104601e3d66f008059ce8669084f41f0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

MarcoFalke commented:
Upgrade to Tested ACK for libfuzzer and honggfuzz

MarcoFalke commented on ci/test/00_setup_env_native_qt5.sh:
```diff
@@ -16,4 +16,5 @@ export RUN_UNIT_TESTS_SEQUENTIAL="true"
 export PREVIOUS_RELEASES_TO_DOWNLOAD="v0.15.2 v0.16.3 v0.17.2 v0.18.1 v0.19.1"
-export BITCOIN_CONFIG="--enable-zmq --with-libs=no --with-gui=qt5 --enable-glibc-back-compat --enable-reduce-exports --enable-debug CFLAGS=\"-g0 -O2 -funsigned-char\" CXXFLAGS=\"-g0 -O2 -funsigned-char\" --with-boost-process"
+export BITCOIN_CONFIG="--enable-zmq --with-libs=no --with-gui=qt5 --enable-glibc-back-compat --enable-reduce-exports
+--enable-debug --disable-fuzz-binary  CFLAGS=\"-g0 -O2 -funsigned-char\" CXXFLAGS=\"-g0 -O2 -funsigned-char\" --with-boost-process"
```
feel free to follow up with pull requests to remove the temporary disable-fuzz-binary

MarcoFalke commented on ci/test/00_setup_env_win64.sh:
```diff
@@ -13,7 +13,7 @@ export PACKAGES="python3 nsis g++-mingw-w64-x86-64 wine-binfmt wine64 file"
 export RUN_SECURITY_TESTS="true"
 export GOAL="deploy"
-export BITCOIN_CONFIG="--enable-reduce-exports --disable-gui-tests --without-boost-process"
+export BITCOIN_CONFIG="--enable-reduce-exports --disable-fuzz-binary --disable-gui-tests --without-boost-process"
```
same

maflcko reviewed (commented):
left some ideas for follow-ups

hebasto commented:
After merging this PR Windows cross compiling fails. Fixed in #21115.
